### PR TITLE
Implement stream.Readable.pipe() for iotjs

### PR DESCRIPTION
**Pull request by pmarkee**
IoT.js-DCO-1.0-Signed-off-by: Peter Marki marpeter@inf.u-szeged.hu

### CODE REVIEW

File: src/js/stream_readable.js [hunk 1]
```diff
@@ -138,6 +139,33 @@ Readable.prototype.push = function(chunk, encoding) {
...
+  options = options || {'end': true};
+
+  var state = this._readableState;
+
+  this.on('readable', function() {
+    this.resume();
+  });
+
+  this.on('data', function(data) {
+    destination.write(data, function() {});
```
Comment (zherczeg) [inline src/js/stream_readable.js]: do we need this `function() {}` ?

Comment (pmarkee) [reply src/js/stream_readable.js]: We don't, I left it there by accident.

File: src/js/stream_readable.js [hunk 2]
```diff
@@ -138,6 +139,76 @@ Readable.prototype.push = function(chunk, encoding) {
 };
 
 
+Readable.prototype.pipe = function(destination, options) {
+  if (!(destination instanceof Writable || isDuplex(destination))) {
```
Comment (rtakacs) [inline src/js/stream_readable.js]: I think it's enough to check if `destination` is readable:

```javascript
if (destination instanceof Readable) {
    throw new TypeError('pipe excepts writable stream argument');
}
```
In this case the `isDuplex` function can be removed.

File: test/run_pass/test_stream_pipe.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+var Duplex = stream.Duplex;
+var Readable = stream.Readable;
+var Writable = stream.Writable;
+
+var readable = new Readable();
+
+var msg = [];
+var resp = [];
+
+// 1
```
Comment (rtakacs) [inline test/run_pass/test_stream_pipe.js]: It's better to use a more informative comment. E.g. `// Step 1` or `// Test 1` or `// Testcase 1`

File: src/js/stream_readable.js [hunk 4]
```diff
@@ -138,6 +139,76 @@ Readable.prototype.push = function(chunk, encoding) {
...
+Readable.prototype.unpipe = function(destination) {
+  if (destination === undefined) {
+    this.removeAllListeners();
+    this._piped = [];
+    this._piped_listeners = [];
+    return;
+  }
+
+  var idx = this._piped.indexOf(destination);
+  if (idx == -1) {
```
Comment (rtakacs) [inline src/js/stream_readable.js]: Please use strict equality operator.

File: test/run_pass/test_stream_pipe.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+dest6._readyToWrite();
+
+readable2.push(msg[2]);
+readable2.pipe(dest6);
+readable2.unpipe(dest6);
+
+process.on('exit', function() {
+  assert.equal(msg[0], resp[0], true);
```
Comment (galpeter) [inline test/run_pass/test_stream_pipe.js]: Why is the third argument `true`? The third one should be an assert message if used.

File: test/run_pass/test_stream_pipe.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+msg.push('Hello');
+
+dest1._write = function(chunk, callback, onwrite) {
+  resp.push(chunk.toString());
+}
+
+dest1._readyToWrite();
+
+readable.pipe(dest1);
```
Comment (galpeter) [inline test/run_pass/test_stream_pipe.js]: We should also add test case when the source and destination is a file and when it is a network socket.

File: test/run_pass/test_stream_pipe.js (hunk 5 above)
Comment (galpeter) [reply test/run_pass/test_stream_pipe.js]: Based on these assert's I'm not sure what are we testing? Can we have some comments, assert messages?

File: src/js/stream_readable.js [hunk 7]
```diff
@@ -138,6 +139,76 @@ Readable.prototype.push = function(chunk, encoding) {
...
+  options = options || {'end': true};
+
+  var listeners = {
+    'readableListener': readableListener.bind(this),
+    'dataListener': dataListener.bind(destination),
+    'endListener': endListener.bind(destination),
+  };
+
+  this.on('readable', listeners.readableListener);
+  this.on('data', listeners.dataListener);
```
Comment (galpeter) [inline src/js/stream_readable.js]: What will happen if there is a stream error? (source/destination)

File: src/js/stream_readable.js (hunk 2 above)
Comment (rtakacs) [reply src/js/stream_readable.js]: Yes, You are right.

File: test/run_pass/test_stream_pipe.js (hunk 5 above)
Comment (pmarkee) [reply test/run_pass/test_stream_pipe.js]: My bad, I knew wrong what the third parameter has to be. I will write some more informative assert messages.

File: src/js/stream_readable.js [hunk 8]
```diff
@@ -138,6 +139,76 @@ Readable.prototype.push = function(chunk, encoding) {
 };
 
 
+Readable.prototype.pipe = function(destination, options) {
```
[the remark this answers is not in the document]
Comment (pmarkee) [reply src/js/stream_readable.js]: According to the node.js documentation, it has to be possible to attach multiple destination streams to the same Readable, and it works here too. I will add a test case for this to `test_stream_pipe.js`.

File: src/js/stream_readable.js (hunk 7 above)
Comment (galpeter) [reply src/js/stream_readable.js]: If the node.js does not state anything else, then it's ok. I'm a bit against adding extra options.

File: src/js/stream_readable.js [hunk 9]
```diff
@@ -138,6 +139,76 @@ Readable.prototype.push = function(chunk, encoding) {
...
+
+  return destination;
+};
+
+
+Readable.prototype.unpipe = function(destination) {
+  if (destination === undefined) {
+    this.removeAllListeners();
+    this._piped = [];
+    this._piped_listeners = [];
```
Comment (zherczeg) [inline src/js/stream_readable.js]: These two could be set to `undefined` since the array is created if needed.

File: src/js/stream_readable.js [hunk 10]
```diff
@@ -138,6 +139,76 @@ Readable.prototype.push = function(chunk, encoding) {
...
+    throw new TypeError('pipe excepts stream.Writable or' +
+                        ' stream.Duplex as argument');
+  }
+
+  options = options || {'end': true};
+
+  var listeners = {
+    'readableListener': readableListener.bind(this),
+    'dataListener': dataListener.bind(destination),
+    'endListener': endListener.bind(destination),
```
Comment (zherczeg) [inline src/js/stream_readable.js]: No need quote around the property names.

**Comment by pmarkee** [comment]
Thanks @zherczeg , I have updated the PR.

**Pull request merged**